Userland/Libraries/LibWeb/HTML/FormControlInfrastructure.cpp

```diff
@@ -40,4 +46,140 @@ WebIDL::ExceptionOr<Entry> create_entry(JS::Realm& realm, DeprecatedString const
     };
 }
 
+// https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set
+WebIDL::ExceptionOr<Optional<HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>>>> construct_entry_list(JS::Realm& realm, HTMLFormElement& form)
+{
+    // 1. If form's constructing entry list is true, then return null.
+    if (form.constructing_entry_list())
+        return Optional<HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>>> {};
+
+    // 2. Set form's constructing entry list to true.
+    form.set_constructing_entry_list(true);
+    ScopeGuard scope_guard([&] {
+        form.set_constructing_entry_list(false);
+    });
+
+    // 3. Let controls be a list of all the submittable elements whose form owner is form, in tree order.
+    auto controls = TRY_OR_THROW_OOM(realm.vm(), form.get_submittable_elements());
+
+    // 4. Let entry list be a new empty entry list.
+    HashMap<DeprecatedString, Vector<Variant<JS::Handle<FileAPI::File>, DeprecatedString>>> entry_list = {};
+
+    // 5. For each element field in controls, in tree order:
+    for (auto control : controls) {
```

Comment:
Should these `control`s be copied? (This might be fine because they seem to be pointers)

Reply:
in that case it should be `auto*`